test/Maps/MapWriter.test.cpp

```diff
@@ -0,0 +1,47 @@
+#include "Maps/MapWriter.h"
+#include "Streams/MemoryWriter.h"
+#include "XFile.h"
+#include <gtest/gtest.h>
+#include <vector>
+#include <stdexcept>
+
+TEST(MapWriter, EmptyMapData)
+{
+	MapData mapData;
+	
+	// Write to File
+	const std::string testFilename("maps/data/test.map");
+	EXPECT_THROW(MapWriter::Write(testFilename, mapData), std::runtime_error);
+	XFile::DeletePath(testFilename);
+
+	// Write to Memory
+	std::vector<char> buffer;
+	buffer.resize(1000);
+	Stream::MemoryWriter memoryWriter(buffer.data(), buffer.size() * sizeof(decltype(buffer)::value_type));
+	EXPECT_THROW(MapWriter::Write(memoryWriter, mapData), std::runtime_error);
+}
+
+TEST(MapWriter, BlankFilename)
+{
+	MapData mapData;
+
+	EXPECT_THROW(MapWriter::Write("", mapData), std::runtime_error);
+}
+
+TEST(MapWriter, VersionTag) {
+	const std::string testFilename("maps/data/test.map");
+	MapData mapData;
+	const int minVersionTag = 0x1010;
+	
+	// The following 2 tests will require a legitimately constructed Map before running.
+	//mapData.header.versionTag = minVersionTag;
+	//EXPECT_NO_THROW(MapWriter::Write(testFilename, mapData));
+
+	//mapData.header.versionTag = minVersionTag + 1;
+	//EXPECT_NO_THROW(MapWriter::Write(testFilename, mapData));
+
+	mapData.header.versionTag = minVersionTag - 1;
+	EXPECT_THROW(MapWriter::Write(testFilename, mapData), std::runtime_error);
```

Comment:
Hmm. Do we want this to fail? This certainly seems worth of a warning, though I'm not sure it constitutes failure.

Perhaps more to the point, should the user even be specifying the version number? The version number is baked into the file format, and that's controlled by the file writing code. Perhaps this value should be auto set.

Reply:
I was approaching MapWriter as having to produce a valid map for consumption with Outpost 2 to be successful. In this case, failure is okay. 

I think baking the version number or not depends on what we want the MapData class to represent. Currently, the goal of MapData is to closely mirror the structure of a Map File on disk with the addition of C++11 containers and small helper functions. It is more of a structure than a class. Eliminating the version number from MapData would diverge from this goal since it is represented on disk.

How do we want to define a valid MapData struct? If we want to say that a valid map file may not be consumable by Outpost 2, then we can eliminate the check.

We could say a valid map file does not contain any logical errors. In this case, I think we would check:
 - Does the number of tiles reported in the header match the recorded number of tiles
 - Does the number of tilesets reported in the header match the recorded number of tilsetSources
 - Can the width of the map be represented by a base 2 logarithm
 - Is the clip rectangle within the valid bounds for the map
 - Is the TerainType::tileRange within the range of valid tile definitions
 - Is the TerrainType::bulldozedTileIndex contained within the TerrainType::tileRange
 - Are the TerrainType::wall indices contained within the TerrainType::tileRange
 - Is the first index for lava tiles contained within the TerrainType::tileRange
 - Is the maximum number of tile definitions allowed Outpost 2? (Maybe this one is allowed in case of extension???)
 - etc

However, this will not necessarily allow the map to work with Outpost 2 because it isn't enforcing the hardcoded values required in TerrainType to work with things like established lava animations, bulldozing, etc. 

Probably more on this later.

-Brett

Reply:
I would argue `MapData` has already diverged from being a simple struct. In particular, it contains STL containers, which don't map directly to disk. You'll notice that while `MapHeader` is a packed struct, `MapData` is not. Further, `MapData` contains a private member function. I think we should make `MapData` a class. The file reading code can become a static factory method. The writing code can become regular member functions.

Of course that's too big of a refactor for this branch. We should deal with that at another time.

As for the MapHeader, perhaps that should become a contained struct, and used as a local variable of the read/write methods. The vector sizes imply some fields. Ones that aren't implied can be stored in a smaller struct, or as direct fields.

The version tag is used in about 3 different places, only one of which is in the header. Hence the version tag is best associated with MapData, rather than MapHeader. That makes using it from MapHeader awkward if MapHeader remains an independent struct in its own file.

----

As for the in-memory format of MapData, I don't think it needs to strictly conform to Outpost 2 limitations. Rather, that should be a separate check. I think some sort of verify method makes sense, but I don't think it should verify stored size fields against computed vector sizes. Rather I think it should initialize a local variable struct with the sizes of those vectors. The verify method is then left mostly with bounds checking, such as making sure the number of tiles and map dimensions fit within what can be represented by the Outpost 2 saved game file format.

I think you did a pretty good job listing potential error conditions. I would want to modify the ones that verify header fields against vector sizes though.